Add all the Jetpack Atomic E2E Users & CI

dpasque opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

## Proposed Changes

This adds everything we need to start running Jetpack E2E tests against the special atomic sites during atomic Jetpack deploys.

To do this, this PR includes:

- [x] Adding all the users to the secrets.
- [x] Adding and supporting the new `ATOMIC_VARIATION`
- [x] Adding a Kotlin library function for making a test running step
- [x] Builds a new Atomic TeamCity build that loops through all the variations
- [x] Supporting the new `RUN_ID` and `RETRY_COUNT` env vars in suite and artifact titles
- [x] Updating existing e2e builds to use the `RETRY_COUNT` variable.

To control the scope and minimize regression testing, I only used the new Kotlin library function for the run tests step for the Jetpack deployment related builds.

## Testing Instructions

There's a few different pieces to test in different ways!

First, you can locally test the new ENV variable handling. With the right combinations, you should see the new Jetpack Atomic users in the run. You should also see updated test suite and artifact names. For example, consider the following local command...
`TEST_ON_ATOMIC=true JETPACK_TARGET=wpcom-deployment ATOMIC_VARIATION=php-old RUN_ID='Atomic: php-old' RETRY_COUNT=1  yarn jest specs/blocks/blocks__jetpack-earn.ts`
... you should see:
- The new Jetpack user for old PHP sites as the test user.
- A new suite name: `Blocks: Jetpack Earn (Atomic: php-old)`
- On failure, an artifact name that has both the `RUN_ID` and the `RETRY_COUNT` included!

Second, post-merge, there will be a new build for Jetpack Atomic deployment testing. It should run tests against all the different atomic variations. Again, can't test until merge.

Finally, all existing E2E builds should work as they did! Namely, the Jetpack Simple build should still run correctly, and all builds, when they have a retry, should use the retry correct in the artifact name. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Commit:
Add all the jetpack e2e atomic users and ENV configuration

This enables testing against various atomic sites used during Jetpack deployments

Commit:
Add a re-usable library function for building e2e teamcity build steps

Commit:
Create the Jetpack Atomic E2E build

worldomonation commented on packages/calypso-e2e/src/secrets/secrets-manager.ts:
```diff
@@ -220,6 +228,27 @@ export class SecretsManager {
 				appleLoginUser: {
 					...fakeAccount,
 				},
+				jetpackAtomicDefaultUser: {
+					...fakeFullAccount,
+				},
+				jetpackAtomicPhpOldUser: {
+					...fakeFullAccount,
+				},
+				jetpackAtomicPhpNewUser: {
+					...fakeFullAccount,
+				},
+				jetpackAtomicEcommPlanUser: {
+					...fakeFullAccount,
+				},
+				jetpackAtomicPrivateUser: {
+					...fakeFullAccount,
+				},
+				jetpackAtomicWpBetaUser: {
+					...fakeFullAccount,
+				},
+				jetpackAtomicWpPreviousUser: {
+					...fakeFullAccount,
+				},
```
**thought:non-blocking** - I'll preface with "I know just enough TypeScript to make wil suggestions".

Is there a way to genericize this check? Having to make changes in three different places is error-prone and tedious.

dpasque replied:
> Is there a way to genericize this check? Having to make changes in three different places is error-prone and tedious.

Heh, I'm totally there with you! 😆 

Yeah, there are ways, but they're actually not as easy or slick as one might hope... The problem is how types are handled in TypeScript -- they're entirely transpile-time and not runtime. As a result, it's kind of infamously annoying to try to do complicated runtime type validation (which is what this is) in TypeScript.

There are some libraries that help do that. Or you can also take care of this by writing recursive functions that crawl through every property and check it against the basic JavaScript types. Or, since this is JSON, we could actually maintain a JSON schema file and then just do a validation against that schema! In all these options, it's kind of a tradeoff of verbosity vs complexity. 

All that to say... I'm definitely interested in revisiting this, but I think we should push that out of the scope of this PR! 😁

noahtallen replied:
The main thing I could think of is turning it into a function which returns `fakeFullAccount` when you want to get one of these known users. Or generate the object with array.reduce over the known account slugs or something like that

(not a big deal obviously!)

dpasque replied:
> Or generate the object with array.reduce over the known account slugs or something like that

Haha, yeah, this would be a relatively easy win and read a lot nicer, but right now the shapes of many of our users are all a little different. 🤦 I need to do a another clean-up or standardization of our secrets file again sometime soon...

Commit:
Add and support RUN_ID and RETRY_COUNT ENV vars

This allow a new way of distinguishing e2e run variations and retries.
This includes support for both suite AND artifact names.

worldomonation commented on test/e2e/specs/blocks/shared/block-smoke-testing.ts:
```diff
@@ -19,7 +20,7 @@ declare const browser: Browser;
  */
 export function createBlockTests( specName: string, blockFlows: BlockFlow[] ): void {
-	describe( specName, function () {
+	describe( DataHelper.createSuiteTitle( specName ), function () {
```
The suite title is back, baby

dpasque replied:
Hahaha, I almost made this _exact_ comment! 😆

worldomonation commented on packages/calypso-e2e/src/types/env-variables.types.ts:
```diff
@@ -8,6 +8,14 @@ export type EnvVariables = {
 export type ViewportName = string;
 export type TestLocales = string[];
 export type JetpackTarget = 'remote-site' | 'wpcom-production' | 'wpcom-deployment';
+export type AtomicVariation =
+	| 'default'
+	| 'php-old'
+	| 'php-new'
+	| 'wp-beta'
+	| 'wp-previous'
+	| 'private'
+	| 'ecomm-plan';
```
**question:non-blocking** - if I'm understanding correctly, we want to eventually run the set of E2E against all these AT environments?

dpasque replied:
Yep! 🚀 

When this PR merges, we'll actually already be doing it for the Jetpack Atomic tests! 😁 But we could extend this to our Gutenberg tests in the future too, which could be pretty rad.

dpasque commented:
I'm going to go ahead and merge this now so we can iterate on this new atomic build and start pushing other dependent pieces of the Jetpack testing CI along 👍
